### PR TITLE
NEW: Add WSI and Tile patch-level prediction functionality

### PR DESCRIPTION
**Pull request by simongraham**
Tile and WSI patch-level prediction.

Add functionality to enable the above. Major changes:

- [x] Add WSIPatchDataset class to deal with tiles and WSIs
- [x] Refactor `CNNPatchPredictor` so that dataset class is now generated on the fly
- [x] `predict` method of `CNNPatchPredictor` takes `mode` as an argument. Internally, this will then generate the correct dataset object.  
- [x] Update `WSIReader` so that we have an option to return the coordinates without needing to read the patches, like is done within the `save_tiles` method.
- [x] Include a masking function to coordinates in `PatchExtractor`
- [x] Include functionality to merge patch predictions to output a 2D prediction map.
- [x] Testing

### CODE REVIEW

File: tiatoolbox/models/classification/patch_predictor.py [hunk 1]
```diff
@@ -266,128 +349,331 @@ def predict(
...
 
-        model = misc.model_to(on_gpu=on_gpu, model=self.model)
+        # use external for testing
+        model = misc.model_to(on_gpu, self.model)
 
-        all_output = {}
-        predictions_output = []
-        probabilities_output = []
-        labels_output = []
+        cum_output = {
```
Comment (shaneahmed) [inline tiatoolbox/models/classification/patch_predictor.py:360]: can we call this `cummulative_output`

Comment (vqdang) [reply tiatoolbox/models/classification/patch_predictor.py:360]: previously i used that, then I think it becomes too long later

File: tests/test_utils.py [hunk 2]
```diff
@@ -993,3 +993,18 @@ def test_model_to():
     model = torch_models.resnet18()
     model = misc.model_to(on_gpu=False, model=model)
     assert isinstance(model, nn.Module)
+
+
+def test_save_json():
+    """Test for placing model on device."""
```
Comment (shaneahmed) [inline tests/test_utils.py:999]: Is this test for model or model output?

File: tiatoolbox/models/dataset/abc.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    def __init__(self, preproc_func=None):
+        super().__init__()
+        self.set_preproc_func(preproc_func)
+        self.data_is_npy_alike = False
+        self.input_list = []
+        self.label_list = []
+
+    def check_input_integrity(self, mode):
+        """Perform check on the input to make sure it is valid."""
```
Comment (shaneahmed) [inline tiatoolbox/models/dataset/abc.py:53]: it is valid image or array?
please can you explain in the docs how you are checking validity?

File: tiatoolbox/models/dataset/abc.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                    self.data_is_npy_alike = False
+                else:
+                    shape_list = [v.shape for v in self.input_list]
+                    self.data_is_npy_alike = True
+
+                if any(len(v) != 3 for v in shape_list):
+                    raise ValueError("Each sample must be an array of the form HWC.")
+
+                max_shape = np.max(shape_list, axis=0)
+                # How will this behave for mixed channel ?
```
Comment (shaneahmed) [inline tiatoolbox/models/dataset/abc.py:91]: Is this a question or a comment?

File: tiatoolbox/models/dataset/abc.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    @staticmethod
+    def load_img(path):
+        """Load an image from a provided path.
+
+        Args:
+            path (str): Path to an image file.
+
+        """
+        path = pathlib.Path(path)
```
Comment (shaneahmed) [inline tiatoolbox/models/dataset/abc.py:128]: can we not merge npy in misc.imread?

File: tiatoolbox/models/dataset/info.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    def label_name(self):
+        raise NotImplementedError
+
+
+class KatherPatchDataset(ABCDatasetInfo):
+    """Define a class for holding the Kather dataset information.
+
+    Args:
+        save_dir_path (str or None): Path to directory containing the Kather dataset,
+                 assumed to be as is after extracted. If the argument is `None`,
```
Comment (shaneahmed) [inline tiatoolbox/models/dataset/info.py:64]: please make sure the indentation is consistent at least in the docstring for the same function/class.

File: tiatoolbox/utils/misc.py [hunk 7]
```diff
@@ -19,25 +19,28 @@
...
+
+from tiatoolbox import rcParam
+from tiatoolbox.utils.exceptions import FileNotSupported
 
 
 def split_path_name_ext(full_path):
     """Split path of a file to directory path, file name and extension.
 
-    Args:
+    Args:x
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:43]: Please can you check this docstring? I think it is not correct.

File: tiatoolbox/utils/misc.py [hunk 8]
```diff
@@ -567,6 +570,25 @@ def unzip_data(zip_path, save_path, del_zip=True):
         os.remove(zip_path)
 
 
+def save_json(output, output_path):
+    """Convert output to a format supported by json.dumps.
+
+    Args:
+        output (dict): Output dictionary to save.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:577]: please don't use `output` as variable name for input argument. You can call it `in_dict` or `input_dict`.

File: tiatoolbox/utils/misc.py [hunk 9]
```diff
@@ -567,6 +570,25 @@ def unzip_data(zip_path, save_path, del_zip=True):
         os.remove(zip_path)
 
 
+def save_json(output, output_path):
+    """Convert output to a format supported by json.dumps.
```
Comment (shaneahmed) [inline tiatoolbox/utils/misc.py:574]: `saves python `dict` in a json file.`

File: tiatoolbox/models/classification/patch_predictor.py [hunk 10]
```diff
@@ -266,128 +345,329 @@ def predict(
...
+                If multiple image tiles or whole-slide images are provided as input,
+                then results are saved to `save_dir` and a dictionary indicating save
+                location for each input is return.
+
+                The dict has following format:
+                - img_path: path of the input image.
+                    - raw: path to save location for raw prediction.
+                    - merged: path to .npy contain merged predictions.
+
+
```
Comment (shaneahmed) [inline tiatoolbox/models/classification/patch_predictor.py:466]: any examples?

File: tiatoolbox/models/classification/patch_predictor.py [hunk 11]
```diff
@@ -266,128 +345,329 @@ def predict(
...
+              Kather 100k dataset.
+            - googlenet-kather100k: googlenet backbone trained on Kather 100k dataset.
 
             By default, the corresponding pretrained weights will also be downloaded.
             However, you can override with your own set of weights via the
             `pretrained_weight` argument. Argument is case insensitive.
 
         pretrained_weight (str): Path to the weight of the
-            corresponding `predefined_model`.
+            corresponding `pretrained_model`.
```
Comment (shaneahmed) [inline tiatoolbox/models/classification/patch_predictor.py:628]: Examples?

**Review by shaneahmed — CHANGES_REQUESTED** [review]
Thanks @vqdang . Please can you make the requested changes?

File: tests/test_models.py [hunk 12]
```diff
@@ -1,270 +1,181 @@
...
-        "wide_resnet101_2",
-        "densenet121",
-        "densenet161",
-        "densenet169",
-        "densenet201",
-        "googlenet",
-        "mobilenet_v2",
-        "mobilenet_v3_large",
-        "mobilenet_v3_small",
+        # "resnet18",
```
Comment (simongraham) [inline tests/test_models.py:43]: Why are these commented out?

File: tests/test_models.py [hunk 13]
```diff
@@ -1,270 +1,181 @@
...
-        ModelBase()
-    with pytest.raises(NotImplementedError):
-        ModelBase.infer_batch(1, 2, 3)
-
-    # without providing any model
-    with pytest.raises(ValueError, match=r"Must provide.*"):
-        CNNPatchPredictor()
+def test_DatasetInfo():
+    """Test for kather patch dataset."""
+    # test defining a subclass of dataset info but not defining
```
Comment (simongraham) [inline tests/test_models.py:73]: I'm actually no longer sure whether it makes sense to add this Kather patch dataset now - this is a different dataset to the kather 100k that we used. Let's have it in this PR but reconsider later.

File: tests/test_models.py [hunk 14]
```diff
@@ -433,504 +337,538 @@ def test_PatchDataset_crash():
...
-    _test_outputs_api1(
-        dataset,
-        "resnet50-kather100k",
-        return_probabilities=True,
-        return_labels=True,
-        probabilities_check=probabilities_check,
-        predictions_check=predictions_check,
+def test_patch_predictor_api(_sample_patch1, _sample_patch2):
+    """Helper function to get the model output using API 1."""
+    # must wrap or sthg stupid happens
```
Comment (simongraham) [inline tests/test_models.py:547]: something stupid? too informal - please give more details

**Review by simongraham — COMMENTED** [review]
Overall it looks good to me. Major difference from when I was working on it is the IO class. A few minor comments to address.

File: tiatoolbox/models/classification/patch_predictor.py [hunk 15]
```diff
@@ -206,20 +202,107 @@ def __init__(
...
+        output,
+        resolution=None,
+        units=None,
+    ):
+        """Merge patch-level predictions to form a 2-dimensional prediction map.
+
+        Args:
+            img (:obj:`str` or :obj:`pathlib.Path` or :class:`numpy.ndarray`):
+                A HWC image or a path to WSI.
+            output_list (list): List of outputs generated by the model.
```
Comment (mostafajahanifar) [inline tiatoolbox/models/classification/patch_predictor.py:238]: it seems that the right argument here is a single `output` not `output_list`. Consider correcting the docstring.

File: tiatoolbox/models/classification/patch_predictor.py [hunk 16]
```diff
@@ -206,20 +202,107 @@ def __init__(
...
+            predictions = output["probabilities"]
+            num_class = np.array(predictions[0]).shape[0]
+            denominator = np.zeros(canvas_shape)
+            output = np.zeros(list(canvas_shape) + [num_class], dtype=np.float32)
+
+        for idx, bound in enumerate(coordinates):
+            prediction = predictions[idx]
+            # assumed to be in XY
+            # top-left for output placement
+            tl = (np.array(bound[:2]) * fx + 0.5).astype(np.int32)
```
Comment (mostafajahanifar) [inline tiatoolbox/models/classification/patch_predictor.py:287]: doesn't `np.floor()` do the same functionality? if yes, it might be better to consider rewriting this line using `np.floor` to keep the code easier to understand.

File: tiatoolbox/models/classification/patch_predictor.py [hunk 17]
```diff
@@ -266,128 +345,329 @@ def predict(
...
+                    - raw: path to save location for raw prediction.
+                    - merged: path to .npy contain merged predictions.
+
+
+        """
+        if mode not in ["patch", "wsi", "tile"]:
+            raise ValueError(
+                "%s is not a valid mode. Use either `patch`, `tile` or `wsi`" % mode
+            )
+        if label_list is not None and mode == "patch":
```
Comment (mostafajahanifar) [inline tiatoolbox/models/classification/patch_predictor.py:472]: should we also check if the length of `img_list`, `label_list`, and `mask_list` are equal?

**Review by mostafajahanifar — APPROVED** [review]

File: tests/test_models.py (hunk 14 above)
Comment (vqdang) [reply tests/test_models.py:547]: For some reason the str going from the dict isnt considered as a path on local machine when running the test. Hence all the input paths to the test are wrapped again via pathlib.Path. It took me awhile to find out back then.

File: tests/test_utils.py (hunk 2 above)
Comment (shaneahmed) [reply tests/test_utils.py:999]: Addressed in this commit https://github.com/TIA-Lab/tiatoolbox/pull/109/commits/0e5e72498ed711e569a359ff33dfa746ba73e6e1

File: tiatoolbox/utils/misc.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/utils/misc.py:574]: still need to clarify the docstring.

File: tiatoolbox/models/dataset/abc.py (hunk 4 above)
Comment (vqdang) [reply tiatoolbox/models/dataset/abc.py:91]: My rambling when codeing, removed now. :)

File: tests/test_models.py (hunk 12 above)
Comment (vqdang) [reply tests/test_models.py]: Uncommented now, likely I locally tested sthg and forgot to add them back.

File: tiatoolbox/models/dataset/abc.py [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+        preproc_func: Preprocessing function used to transform the input data. If
+        supplied, then torch.Compose will be used on the input preproc_list.
+        preproc_list is a list of torchvision transforms for preprocessing the image.
+        The transforms will be applied in the order that they are given in the list.
+        https://pytorch.org/vision/stable/transforms.html.
+
+    """
+
+    def __init__(self, preproc_func=None):
```
Comment (John-P) [inline tiatoolbox/models/dataset/abc.py:46]: Should this init have arguments ofr input_list and label_list?

File: tiatoolbox/models/dataset/classification.py [hunk 19]
```diff
@@ -65,254 +73,266 @@ def predefined_preproc_func(dataset_name):
...
     Attributes:
-        return_labels (bool, False): `__getitem__` will return both the img and
-        its label.
-                If `label_list` is `None`, `None` is returned
+        input_list: Either a list of patches, where each patch is a ndarray or a list of
+        valid path with its extension be (".jpg", ".jpeg", ".tif", ".tiff", ".png")
+        pointing to an image.
+
+        label_list: List of label for sample at the same index in `input_list` .
+        Default is `None`.
```
Comment (John-P) [inline tiatoolbox/models/dataset/classification.py:94]: Suggest removing _list from the arument names and just making them plural instead. The fact that it should be a list should be clarified by type annotations and the docstring.

'input_list' -> 'inputs' or 'patches'
'label_list' -> 'labels'

File: tiatoolbox/models/dataset/classification.py [hunk 20]
```diff
@@ -65,254 +73,266 @@ def predefined_preproc_func(dataset_name):
...
-        path = pathlib.Path(path)
-        if path.suffix == ".npy":
-            patch = np.load(path)
-        elif path.suffix in (".jpg", ".jpeg", ".tif", ".tiff", ".png"):
-            patch = imread(path)
-        else:
-            raise ValueError("Can not load data of `%s`" % path.suffix)
-        return patch
+    Examples:
+        >>> from tiatoolbox.models.data import Patch_Dataset
```
Comment (John-P) [inline tiatoolbox/models/dataset/classification.py:103]: Remove underscore in `Patch_Dataset`

File: tiatoolbox/models/dataset/classification.py (hunk 19 above)
Comment (John-P) [reply tiatoolbox/models/dataset/classification.py:94]: Also could possibly use x and y similarly to sklearn / model fitting math convention.

**Comment by John-P** [comment]
Moslty just variable name suggestions and some confusion of the IOState class.

File: tiatoolbox/models/classification/patch_predictor.py [hunk 21]
```diff
@@ -266,128 +372,345 @@ def predict(
...
+
+    def predict(
+        self,
+        img_list,
+        mask_list=None,
+        label_list=None,
+        mode="patch",
+        return_probabilities=False,
+        return_labels=False,
+        on_gpu=True,
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/classification/patch_predictor.py:428]: We could masking gpu beforehand via `os.environ` hence why I didnt open up the above option.

File: tiatoolbox/models/classification/patch_predictor.py [hunk 22]
```diff
@@ -20,19 +20,45 @@
...
-from tiatoolbox.models.classification.pretrained_info import _pretrained_model
+from tiatoolbox.utils.misc import (
+    download_data,
+    get_pretrained_model_info,
+    save_dict_to_json,
+)
+from tiatoolbox.wsicore.wsireader import VirtualWSIReader, get_wsireader
+
+
+class _IOStatePatchPredictor(IOStateBase):
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/classification/patch_predictor.py:48]: Will rename to IOConfig as 2021-07-16 meeting.

File: tiatoolbox/wsicore/wsireader.py [hunk 23]
```diff
@@ -185,7 +185,7 @@ def level_to_downsample(x):
             base_scale = 1
             resolution = 1 / resolution
         else:
-            raise ValueError("Invalid units")
+            raise ValueError("Invalid units `%s`." % units)
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/wsicore/wsireader.py:188]: I mentioned this briefly in the chat, I have been cleaning it up but I may miss some such as these. Thanks for spotting it,

File: tiatoolbox/utils/misc.py [hunk 24]
```diff
@@ -567,6 +576,29 @@ def unzip_data(zip_path, save_path, del_zip=True):
         os.remove(zip_path)
 
 
+def save_dict_to_json(input_dict, save_path):
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/utils/misc.py:579]: Renamed to `save_as_json` and allow list as input. However I'm worried this will create misusage later on because this is very primitive because it only converts np.ndarray, other type is passed through

File: tiatoolbox/models/dataset/abc.py [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+            raise ValueError(f"Can not load data of `{path.suffix}`")
+        return patch
+
+    @staticmethod
+    def preproc(image):
+        """Define the pre-processing of this class of loader."""
+        return image
+
+    @property
+    def preproc_func(self):
```
Comment (vqdang) [inline tiatoolbox/models/dataset/abc.py:147]: @John-P

File: tiatoolbox/models/dataset/info.py [hunk 26]
```diff
@@ -0,0 +1,129 @@
...
+    # We predefine to follow enforcement, actual initialization in init
+    inputs = None
+    labels = None
+    label_name = None
+
+    def __init__(
+        self,
+        save_dir_path=None,
+    ):
+        label_codes = [
```
Comment (John-P) [inline tiatoolbox/models/dataset/info.py:86]: Suggest 'label_name' instead of 'label_codes' as it's easier to guess what it is from the name.

File: tiatoolbox/models/dataset/info.py [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+class ABCDatasetInfo(ABC):
+    """Define an abstract class for holding a dataset information.
+
+    Enforcing such that following attributes must always be defined by the subclass.
+
+    Attributes
+        input_list (list): A list of paths where each path points to a sample image.
+        label_list (list): A list of `int` where each is the label of the sample at
+            the same index.
+        label_name (dict): A dict indicates the possible associate name of each label
```
[the remark this answers is not in the document]
Comment (vqdang) [reply tiatoolbox/models/dataset/info.py]: Copy over from team chat
```
I think you are confused between PatchDataset(torch) and this one. This one is just data class holder, just like sklearn iris. 
>tiatoolbox/models/dataset/info.py
While this other one is for defining torch loading IO etc.
>tiatoolbox/models/dataset/classification.py
>tiatoolbox/models/dataset/abc.py
```

**Review by John-P — APPROVED** [review]

**Review by shaneahmed — APPROVED** [review]
WellDone @vqdang 👏👏👏
